stan/math/prim/meta/holder.hpp

```diff
@@ -0,0 +1,274 @@
+  }
+
+  EIGEN_STRONG_INLINE Scalar& coeffRef(Index row, Index col) {
+    return m_argImpl.coeffRef(row, col);
+  }
+  EIGEN_STRONG_INLINE Scalar& coeffRef(Index index) {
+    return m_argImpl.coeffRef(index);
+  }
+
+  template <int LoadMode, typename PacketType>
+  EIGEN_STRONG_INLINE PacketType packet(Index row, Index col) const {
+    return m_argImpl.template packet<LoadMode, PacketType>(row, col);
+  }
+  template <int LoadMode, typename PacketType>
+  EIGEN_STRONG_INLINE PacketType packet(Index index) const {
+    return m_argImpl.template packet<LoadMode, PacketType>(index);
+  }
+
+  template <int StoreMode, typename PacketType>
+  EIGEN_STRONG_INLINE void writePacket(Index row, Index col,
+                                       const PacketType& x) {
+    return m_argImpl.template writePacket(row, col, x);
+  }
+  template <int StoreMode, typename PacketType>
+  EIGEN_STRONG_INLINE void writePacket(Index index, const PacketType& x) {
+    return m_argImpl.template writePacket(index, x);
+  }
+};
+
+}  // namespace internal
+}  // namespace Eigen
+
+namespace stan {
+namespace math {
+
+/**
+ * Constructs a no-op operation that also holds pointer to some other
+ * expressions, allocated on heap. When the object is destructed those
+ * expressions will be deleted.
+ * @tparam T type of argument expression
+ * @tparam Ptrs types of pointers
+ * @param a argument expression
+ * @param ptrs pointers to objects the constructed object will own.
+ * @return holder operation
+ */
+template <typename T, typename... Ptrs,
+          std::enable_if_t<sizeof...(Ptrs) >= 1>* = nullptr>
+Holder<T, Ptrs...> holder(T&& arg, Ptrs*... pointers) {
+  return Holder<T, Ptrs...>(std::forward<T>(arg), pointers...);
+}
+// trivial case with no pointers constructs no holder object
+template <typename T>
+T holder(T&& arg) {
+  return std::forward<T>(arg);
+}
+
+namespace internal {
+// the function holder_handle_element is also used in holder_cl
+/**
+ * Handles single element (moving rvalues to heap) for construction of
+ * `holder` or `holder_cl` from a functor. For lvalues just sets the `res`
+ * pointer.
+ * @tparam T type of the element
+ * @param a element to handle
+ * @param res resulting pointer to element
+ * @return tuple of pointer allocated on heap (empty).
+ */
+template <typename T>
+auto holder_handle_element(T& a, T*& res) {
+  res = &a;
+  return std::make_tuple();
+}
+
+/**
+ * Handles single element (moving rvalues to heap) for construction of
+ * `holder` or `holder_cl` from a functor. Rvalue is moved to heap and the
+ * pointer to heap memory is assigned to res and returned in a tuple.
+ * @tparam T type of the element
+ * @param a element to handle
+ * @param res resulting pointer to element
+ * @return tuple of pointer allocated on heap (containing single pointer).
+ */
+template <typename T>
+auto holder_handle_element(std::remove_reference_t<T>&& a, T*& res) {
+  res = new T(std::move(a));
+  return std::make_tuple(res);
+}
```

Comment:
I'm not following why the `std::remove_reference_t` here?

Reply:
Also when does the `res` pointer get cleaned up?

Reply:
If you just use `T&&` here could you just use an `enable_if_t<is_rvalue<>>` instead of this?